core/src/main/java/org/infinispan/transaction/xa/XaTransactionTable.java

```diff
@@ -100,32 +104,47 @@ public int getLocalTxCount() {
       return xid2LocalTx.size();
    }
 
-   public int prepare(Xid externalXid) throws XAException {
+   public CompletionStage<Integer> prepare(Xid externalXid) {
       Xid xid = convertXid(externalXid);
       LocalXaTransaction localTransaction = getLocalTransactionAndValidate(xid);
+      if (localTransaction == null) {
+         return CompletableFutures.completedExceptionFuture(new XAException(XAException.XAER_NOTA));
+      }
       return txCoordinator.prepare(localTransaction);
    }
 
-   public void commit(Xid externalXid, boolean isOnePhase) throws XAException {
+   public CompletionStage<Void> commit(Xid externalXid, boolean isOnePhase) {
       Xid xid = convertXid(externalXid);
       LocalXaTransaction localTransaction = getLocalTransactionAndValidate(xid);
-      boolean committedInOnePhase;
+      if (localTransaction == null) {
+         return CompletableFutures.completedExceptionFuture(new XAException(XAException.XAER_NOTA));
+      }
+      CompletionStage<?> prepareStage;
       if (isOnePhase) {
          //isOnePhase being true means that we're the only participant in the distributed transaction and TM does the
          //1PC optimization. We run a 2PC though, as running only 1PC has a high chance of leaving the cluster in
          //inconsistent state.
-         txCoordinator.prepare(localTransaction);
+         prepareStage = txCoordinator.prepare(localTransaction);
+      } else {
+         prepareStage = CompletableFutures.completedNull();
       }
-      committedInOnePhase = txCoordinator.commit(localTransaction, false);
-      forgetSuccessfullyCompletedTransaction(recoveryManager, localTransaction.getXid(), localTransaction,
-            committedInOnePhase);
+      CompletionStage<Boolean> commitStage = prepareStage
+            .thenCompose(ignore -> txCoordinator.commit(localTransaction, false));
```

Comment:
Nitpicking, but I would prefer wrapping before `prepareStage`, or even

      return prepareStage.thenCompose(ignore -> txCoordinator.commit(localTransaction, false))
                         .thenApply(committedInOnePhase -> {

Reply:
Changed to the suggested code.